reckon-core/src/main/groovy/org/ajoberstar/reckon/core/InventoryService.java

```diff
@@ -0,0 +1,214 @@
+
+      Set<ReckonVersion> parallelVersions =
+          taggedVersions
+              .stream()
+              .map(version -> findParallel(walk, headCommit, version, taggedVersions))
+              // TODO Java 9 Optional::stream
+              .flatMap(opt -> opt.isPresent() ? Stream.of(opt.get()) : Stream.empty())
+              .collect(Collectors.toSet());
+
+      Set<ReckonVersion> claimedVersions =
+          taggedVersions.stream().map(TaggedVersion::getVersion).collect(Collectors.toSet());
+
+      return new Inventory(
+          currentVersion,
+          commitsSinceBase,
+          baseRelease.getVersion(),
+          baseVersion.getVersion(),
+          parallelVersions,
+          claimedVersions);
+    } catch (IOException e) {
+      throw new UncheckedIOException(e);
+    }
+  }
+
+  private Set<TaggedVersion> getTaggedVersions(RevWalk walk) throws IOException {
+    Set<TaggedVersion> versions = new HashSet<>();
+
+    for (Ref ref : repo.getRefDatabase().getRefs(Constants.R_TAGS).values()) {
+      Ref tag = repo.peel(ref);
+      ObjectId objectId = tag.getPeeledObjectId();
+      RevCommit commit = walk.parseCommit(objectId);
+      tagParser
+          .apply(tag)
+          .ifPresent(
+              version -> {
+                versions.add(new TaggedVersion(version, commit));
+              });
+    }
+    return versions;
+  }
+
+  private Optional<TaggedVersion> findCurrent(RevCommit head, Stream<TaggedVersion> versions) {
+    return versions
+        .filter(version -> version.getCommit().equals(head))
+        .max(Comparator.comparing(TaggedVersion::getVersion));
+  }
+
+  private TaggedVersion findBase(RevWalk walk, RevCommit head, Stream<TaggedVersion> versions)
+      throws IOException {
+    walk.reset();
+    walk.setRevFilter(RevFilter.ALL);
+    walk.markStart(head);
+
+    Map<RevCommit, List<TaggedVersion>> versionsByCommit =
+        versions.collect(Collectors.groupingBy(TaggedVersion::getCommit));
+
+    Stream.Builder<List<TaggedVersion>> builder = Stream.builder();
+
+    for (RevCommit commit : walk) {
+      List<TaggedVersion> matches = versionsByCommit.get(commit);
+      if (matches != null) {
+        // Parents can't be "nearer". Exclude them to avoid extra walking.
+        for (RevCommit parent : commit.getParents()) {
+          walk.markUninteresting(parent);
+        }
+        builder.accept(matches);
+      }
+    }
+
+    return builder
+        .build()
+        .flatMap(List::stream)
+        .max(Comparator.comparing(TaggedVersion::getVersion))
+        .orElse(new TaggedVersion(ReckonVersion.VERSION_0, null));
+  }
+
+  private Optional<ReckonVersion> findParallel(
+      RevWalk walk, RevCommit head, TaggedVersion candidate, Set<TaggedVersion> versions) {
+    try {
+      walk.reset();
+      walk.setRevFilter(RevFilter.MERGE_BASE);
+      walk.markStart(head);
+      walk.markStart(candidate.getCommit());
+
+      // TODO I don't think this is right
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.ajoberstar.com/coding_rules#rule_key=squid%3AS1135)